src/client/visualizations/heat-map/heat-map.tsx

```diff
@@ -0,0 +1,160 @@
+    return <HeatmapWithTooltip {...this.props} dataset={dataset} />;
+  }
+}
+
+const memoizedFillDatasetWithMissingValues = memoize(fillDatasetWithMissingValues);
+
+const splitToFillOrder = memoize((split: Split): Order<any> => {
+  const sort = split.sort;
+  switch (split.type) {
+    case SplitType.string:
+    case SplitType.number:
+    default:
+      if (sort.direction === SortDirection.ascending) {
+        return orderByValueIncreasing;
+      } else {
+        return orderByValueDecreasing;
+      }
+    case SplitType.time:
+      if (sort.direction === SortDirection.ascending) {
+        return orderByTimeDimensionIncreasing;
+      } else {
+        return orderByTimeDimensionDecreasing;
+      }
+  }
+});
+
+export class UndecoratedHeatmapWithTooltip extends React.Component<VisualizationProps & TooltipProps<RectangleData> & { dataset: Dataset }> {
+  private container: HTMLDivElement | null = null;
+
+  handleRectangleHover = (tooltipData: RectangleData) => {
+    setTimeout(() => {
+      if (!this.container) {
+        return;
+      }
+      const { x, y } = tooltipData;
+      const { top, left } = this.container.getBoundingClientRect();
+
+      this.props.showTooltip({
+        tooltipLeft: x - left,
+        tooltipTop: y - top,
+        tooltipData
+      });
+    }, 0);
+  }
+
+  renderDatum(datum: Datum, concreteSeries: ConcreteSeries): JSXNode {
+    if (!this.props.essence.hasComparison()) {
+      return concreteSeries.formatValue(datum);
+    }
+    const currentValue = concreteSeries.selectValue(datum);
+    const previousValue = concreteSeries.selectValue(datum, SeriesDerivation.PREVIOUS);
+    const formatter = concreteSeries.formatter();
+
+    return <MeasureBubbleContent
+      lowerIsBetter={concreteSeries.measure.lowerIsBetter}
+      formatter={formatter}
+      current={currentValue}
+      previous={previousValue}
+    />;
+  }
+
+  render() {
+    const {
+      tooltipData,
+      tooltipLeft,
+      tooltipTop,
+      tooltipOpen,
+      hideTooltip,
+      essence,
+      dataset
+    } = this.props;
+
+    const { timezone } = essence;
+    const concreteSeries = essence.getConcreteSeries().first();
+    const secondSplit = essence.splits.splits.get(1);
+
+    const preparedDataset = memoizedFillDatasetWithMissingValues(
+      (dataset.data[0][SPLIT] as Dataset),
+      concreteSeries.plywoodKey(),
+      secondSplit.reference,
+      splitToFillOrder(secondSplit),
+      timezone
+    ).data;
+
+    return <div ref={container => this.container = container} className="internals heatmap-container" style={{ maxHeight: this.props.stage.height }}>
+      <LabelledHeatmap
+        dataset={preparedDataset}
+        essence={this.props.essence}
+        onHover={this.handleRectangleHover}
+        onHoverStop={hideTooltip}
+      />
+      {tooltipOpen && (
+        <TooltipWithBounds
+          // set this to random so it correctly updates with parent bounds
+          key={Math.random()}
```

Comment:
😵

Reply:
It's from vx docs. It really does not work properly without it 😂